Reinstated Compression sigil.

keraldi opened the pull request:
- does not compress wooden planks into crafting tables
- searches for reversible 3x3/2x2 recipes with a single material type (searches for 3x3 first)
- caches recipes in generic list recipes<IRecipe> after compressed for the first time
- recipes must be reversible for compression to be allowed
- finally learning for my exams as I should.

keraldi committed:
Reinstated Compression sigil.
- does not compress wooden planks into crafting tables
- searches for reversible 3x3/2x2 recipes with a single material type
- probably has a lot of redundant stuff and looks silly
- uses try/catch (might want to find a different approach, some people scoff at this)
- should probably have spent the night sleeping, I'm taking exams tomorrow and will probably sleep the whole day through.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,28 +4,33 @@
     public ItemStack test(ItemStack[] inv, boolean doDrain, World world) {
         for (ItemStack invStack : inv) {
-            if (invStack.isEmpty()) {
+            if (invStack.isEmpty() || invStack.equals(ItemStack.EMPTY)) {
```
Needless additional check

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,28 +4,33 @@
                 ItemStack stack = getRecipe(invStack, world, i);
-                if (!stack.isEmpty()) {
-                    int threshold = CompressionRegistry.getItemThreshold(invStack);
+                if (!stack.isEmpty() && !invStack.equals(ItemStack.EMPTY) && invStack != ItemStack.EMPTY) {
```
Needless additional checks

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -42,7 +47,7 @@ public int iterateThroughInventory(ItemStack required, int kept, ItemStack[] inv
             i++;
 
-            if (invStack.isEmpty()) {
+            if (invStack == null || invStack.isEmpty() || invStack.equals(ItemStack.EMPTY)) {
```
Needless additional checks

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -88,26 +94,28 @@ public boolean canInteractWith(EntityPlayer player) {
-
-        ItemStack returnStack = StorageBlockCraftingManager.getInstance().findMatchingRecipe(inventory, world);
-        if (returnStack.isEmpty()) {
+        try {
```
try/catch is slow. You should be properly nullchecking the recipe you get back.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -116,8 +124,21 @@ public boolean canInteractWith(EntityPlayer player) {
+            return StorageBlockCraftingManager.getInstance().findMatchingRecipe(inventory, world);
+        }
+
+        try{result = CraftingManager.findMatchingRecipe(inventory, world).getRecipeOutput();
```
Again, try/catch is slow. Properly check your result.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/BaseCompressionHandler.java:
```diff
@@ -50,7 +50,7 @@ public int iterateThroughInventory(ItemStack[] inv, boolean doDrain) {
             i++;
 
-            if (invStack == null) {
+            if (invStack.isEmpty() || invStack.equals(ItemStack.EMPTY) || invStack == ItemStack.EMPTY) {
```
More needless checks

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/CompressionRegistry.java:
```diff
@@ -38,7 +38,7 @@ public static void registerItemThreshold(ItemStack stack, int threshold) {
         for (CompressionHandler handler : compressionRegistry) {
             ItemStack stack = handler.compressInventory(inv, world);
-            if (stack != null) {
+            if (stack != null && stack != ItemStack.EMPTY) {
```
`stack.isEmpty()`

keraldi committed:
Learned how to properly handle the "NullPointerException"-situation.

Commit:
Update BaseCompressionHandler.java

Commit:
Update CompressionRegistry.java

keraldi commented:
Done.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -89,22 +95,21 @@ public boolean canInteractWith(EntityPlayer player) {
                 break;
+        }*/
+        ItemStack returnStack = ItemStack.EMPTY;
+        if(CraftingManager.findMatchingRecipe(inventory, world) != null) {
```
Don't make this query twice. Store your result.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -116,8 +121,21 @@ public boolean canInteractWith(EntityPlayer player) {
+            return StorageBlockCraftingManager.getInstance().findMatchingRecipe(inventory, world);
+        }
+        ItemStack result = ItemStack.EMPTY;
+        if(CraftingManager.findMatchingRecipe(inventory, world) != null){
```
Same here

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -116,8 +121,21 @@ public boolean canInteractWith(EntityPlayer player) {
+        }
+
+        if(isResultStackReversible(result, gridSize, world)){
+            StorageBlockCraftingManager.getInstance().addRecipe(CraftingManager.findMatchingRecipe(inventory, world));
```
Same here...

TehNut reviewed (changes requested):
Delete any commented code. VCS exists for a reason.

keraldi committed:
Removed (almost) all code comments (only a one-liner remains that serves as pointer to a completly commented-out class (StorageBlockCraftingRecipeAssimilator)).
Made methods and variables for long function calls to be more efficient.
Fixed an oversight that resulted in a NullPointerException after removing redundant checks that were made to prevent exactly that.

Rearranged and reformatted code.

keraldi committed:
Merge remote-tracking branch 'origin/compressionv4' into compressionv4

# Conflicts:
#	src/main/java/WayofTime/bloodmagic/compress/CompressionRegistry.java

keraldi committed:
corrected something that could be considered a typo but was an oversight

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,12 +4,82 @@
+        if (stack.isEmpty()) {
+            return false;
+        }
+        InventoryCrafting inventory = new InventoryCrafting(new Container() {
```
This should be stored as a constant. No need to create a new instance each time this is called.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,12 +4,82 @@
+
+    public static ItemStack getRecipe(ItemStack stack, World world, int gridSize) {
+        StorageBlockCraftingManager craftingManagerSB = StorageBlockCraftingManager.getInstance();
+        InventoryCrafting inventory = new InventoryCrafting(new Container() {
```
To go with the above, maybe a `Cache<Point2i, InventoryCrafting>`?

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,12 +4,82 @@
+        return ItemStack.EMPTY;
+    }
+
+    public static boolean has22Recipe(ItemStack stack, World world) {
```
Remove this. Anybody using this is being inefficient since they can just call `get22Recipe(...)` and check if it's empty themselves.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java:
```diff
@@ -4,12 +4,82 @@
+        return getRecipe(stack, world, 2);
+    }
+
+    public static boolean has33Recipe(ItemStack stack, World world) {
```
Remove this. Anybody using this is being inefficient since they can just call `get33Recipe(...)` and check if it's empty themselves.

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/BaseCompressionHandler.java:
```diff
@@ -85,7 +85,9 @@ public int iterateThroughInventory(ItemStack[] inv, boolean doDrain) {
         return needed;
     }
 
+
```
Extra line

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/CompressionHandler.java:
```diff
@@ -12,4 +12,5 @@
      * @return The result of the compression
      */
     public abstract ItemStack compressInventory(ItemStack[] inv, World world);
+
```
Extra line

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/BaseCompressionHandler.java:
```diff
@@ -85,7 +85,9 @@ public int iterateThroughInventory(ItemStack[] inv, boolean doDrain) {
         return this.leftover;
     }
 }
+
```
Extra line

TehNut commented on src/main/java/WayofTime/bloodmagic/compress/CompressionRegistry.java:
```diff
@@ -78,13 +78,15 @@ public static ItemStack compressInventory(ItemStack[] inv, World world) {
 
     public static int getItemThreshold(ItemStack stack) {
-        for (Map.Entry<ItemStack, Integer> entry : thresholdMap.entrySet()) {
+        /*for (Map.Entry<ItemStack, Integer> entry : thresholdMap.entrySet()) {
```
Delete

Commit:
This should be it. An Array should be more efficient but you can correct me if I'm wrong. In either case it does what it is supposed to do.

Commit:
Fixed. Needed a seperate inventory for the reversible check (or clear the previously used one, but then I'd had to repopulate again and that would just have been messy)

Commit:
Forgot one of my lines.

TehNut commented:
Is this tested on a server with multiple players compressing things?

keraldi commented:
Works if multiple players use it simultaneously but there's a dupe that creates i extra compressed items for 8*i +1 re-compressed items.

Commit:
Fix and cleanup.
Could definitely clean more but that should suffice for now.

keraldi commented:
Done. Something about a function with different parameters returning different results.

TehNut commented:
![image](https://user-images.githubusercontent.com/5789439/44632510-b1202b80-a930-11e8-9473-6516828b8084.png)
